src/Server/CertReloader.cpp

```diff
@@ -0,0 +1,99 @@
+#include "CertReloader.h"
+
+namespace DB
+{
+#if USE_SSL
+int cert_reloader_dispatch_set_cert(SSL * ssl, [[maybe_unused]] void * arg)
+{
+    return CertReloader::instance().SetCert(ssl);
+}
+
+int CertReloader::SetCert(SSL * ssl)
+{
+    std::shared_lock lock(lck);
+    SSL_use_certificate(ssl, const_cast<X509 *>(cert->certificate()));
+    SSL_use_RSAPrivateKey(ssl, key->impl()->getRSA());
```

Comment:
Again I have concerns about ownership here.

Reply:
Summarizing what we found:

CertReloader owns cert, key                                                      
                                                                                 
key is a smart pointer to a poco RSAKey (with RSAKeyImpl)                        
reset()'ing key results in  RSAKeyImpl::~RSAKeyImpl                              
RSAKeyImpl::~RSAKeyImpl calls RSA_free                                           
RSA_free decrefs and eventually free()'s the memory                              
                                                                                 
cert is a smart pointer to a poco X509Certificate                                
reset()'ing cert results in X509Certificate::~X509Certificate                    
X509Certificate::~X509Certificate calls X509_free                                
X509_free decrefs and eventually free()'s the memory                             
                                                                                 
callback calls SSL_use_RSAPrivateKey to share *key with openssl                  
SSL_use_RSAPrivateKey does an upref so this *should* be free'd by dtor or when   
the key is no longer referenced.                                                 
                                                                                 
callback calls SSL_use_certificate to share *cert with openssl                   
SSL_use_certificate calls ssl_set_cert(owned ssl->cert, borrowed *cert)          
ssl_set_cert uprefs *cert                                                        
Similar to the key eventually it will be free()'d.                               
                                                                                 
                                                                                 
There is a period of time between when CertReloader reliquishes ownership and    
the memory is free()'d. This means we need to trust ssl to eventually free the   
memory -- if it didn't that would be a bug.                                      
                                                                                 
We can code defensively and kick deadrefs into a queue of sorts and periodically 
troll the list for objects to collect. I think it's extra effort without much    
benefit.                                                                         
                                                                                 
Most people will probably rotate a cert a few days before expiration so the      
worst thing that would happen if this bug exists would be a few hundred KB of    
memory leaked (assuming the certs are properly rotated).

Reply:
We can trust the deref system.